src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs

```diff
@@ -0,0 +1,205 @@
+        }
+
+        // Wrap the function implementation in a lambda orchestrator
+        string orchestratorOutput = OrchestrationRunner.LoadAndRun<object, object>(
+            encodedOrchestratorState,
+            async (orchestrationContext, input) =>
+            {
+                // Set the function input to be the orchestration context wrapped in our own object so that we can
+                // intercept any of the calls and inject our own logic or tracking.
+                FunctionsOrchestrationContext wrapperContext = new(orchestrationContext, functionContext);
+                triggerInputData.Value = wrapperContext;
+
+                // This method will advance to the next middleware and throw if it detects an asynchronous execution.
+                await EnsureSynchronousExecution(functionContext, next, wrapperContext);
+
+                // Set the raw function output as the orchestrator output
+                object? functionOutput = functionContext.GetInvocationResult().Value;
+                return functionOutput;
+            },
+            functionContext.InstanceServices);
+
+        // Send the encoded orchestrator output as the return value seen by the functions host extension
+        functionContext.GetInvocationResult().Value = orchestratorOutput;
+    }
+
+    private static bool IsOrchestrationTrigger(
+        FunctionContext context,
+        [NotNullWhen(true)] out BindingMetadata? orchestrationTriggerBinding)
+    {
+        foreach (BindingMetadata binding in context.FunctionDefinition.InputBindings.Values)
+        {
+            if (string.Equals(binding.Type, "orchestrationTrigger"))
+            {
+                orchestrationTriggerBinding = binding;
+                return true;
+            }
+        }
+
+        orchestrationTriggerBinding = null;
+        return false;
+    }
+
+    private static async Task EnsureSynchronousExecution(
+        FunctionContext functionContext,
+        FunctionExecutionDelegate next,
+        FunctionsOrchestrationContext orchestrationContext)
+    {
+        Task orchestratorTask = next(functionContext);
+        if (!orchestratorTask.IsCompleted && !orchestrationContext.IsAccessed)
+        {
+            // If the middleware returns before the orchestrator function's context object was accessed, then we
+            // know that either some middleware component went async or that the orchestrator function did some illegal
+            // await as its very first action.
+            throw new InvalidOperationException(IllegalAwaitErrorMessage);
+        }
+
+        await orchestratorTask;
+
+        // This will throw if either the orchestrator performed an illegal await or if some middleware ahead of this
+        // one performed some illegal await.
+        orchestrationContext.ThrowIfIllegalAccess();
+    }
+
+    private sealed class FunctionsOrchestrationContext : TaskOrchestrationContext
```

Comment:
This is just a wrapper class for the orchestration context that checks to see if there are any illegal awaits or illegal async callbacks in the user code.